[AIE][Test] Add GPT-2 and core ops implementation

Fangtangtang opened the pull request:
<!--- Copyright Allo authors. All Rights Reserved. -->
<!--- SPDX-License-Identifier: Apache-2.0  -->

## Description ##
This PR provides a GPT-2 implementation for aie backend. 
It also includes independent test cases for key components of the model, such as GEMM and attention score computation.

In addition, this PR fixes a potential bug in `ExternalModule` to avoid naming conflicts between user-defined external modules and external kernel libraries.

### Examples ###
GPT-2 implementation: `tests/dataflow/aie/gpt2/test_gpt2_block.py`

attention score: 
- independent test case: `tests/dataflow/aie/gpt2/test_attn_score.py`
- external kernel: `tests/dataflow/aie/gpt2/attn_score.cc`

GEMM: `tests/dataflow/aie/test_gemm.py`
## Checklist ##

Please make sure to review and check all of these items:
- [x] PR's title starts with a category (e.g. [Bugfix], [IR], [Builder], etc)
- [x] All changes have test coverage (It would be good to provide ~2 different test cases to test the robustness of your code)
- [x] Pass the [formatting check](https://cornell-zhang.github.io/allo/developer/index.html#id1) locally
- [ ] Code is well-documented

chhzh123 commented on allo/backend/experimental/external_kernel.py:
```diff
@@ -45,6 +45,7 @@ def __init__(
         assert self.filename.endswith(
             ".cc"
         ), f"Expected a .cc file, but got: {self.filename}"
+        self.filename = self.filename.removesuffix(".cc") + "_.cc"
```
Any reason for this?

chhzh123 replied:
Adding a comment for this would be better

chhzh123 commented on tests/dataflow/aie/gpt2/attn_score.cc:
```diff
@@ -0,0 +1,52 @@
+
+extern "C" {
+
+void partial_attn_score(float A_in[32][64], float B_in[32][64],
```
What is attention score? `transpose_matmul_with_scale` is much clearer

chhzh123 commented on tests/dataflow/aie/test_gemm.py:
Modify the file name to sth like `test_gemm_temporal_reduction.py`.

chhzh123 commented on tests/dataflow/aie/gpt2/layer_norm.cc:
These two `.cc` files should be inside the `allo` folder? We can construct a library folder for `aie`.

chhzh123 reviewed (commented):
LGTM. Just some file organization and naming issues.

Fangtangtang commented:
Hi, I've fixed some naming issues you mentioned. could you take another look and see if things look better now?

Also, I added an implementation of masked softmax, but it's currently a bit unstable. There’s a strange bug where the result is partially undefined if it’s only run once. However, if it's run multiple times, the final result is correct. It feels like some memory-related bug, and it might not be on our side.
